elki-core-math/src/main/java/elki/math/statistics/tests/mcde/MWPTest.java

```diff
@@ -0,0 +1,191 @@
+   * double l = corrected_rank[0]; double adjusted_rank = corrected_rank[1]; double correction = corrected_rank[2];
+   * <p>
+   * Correspond to one instance of the original data
+   */
+  public <A> MwpIndex[] corrected_ranks(final NumberArrayAdapter<?, A> adapter, final A data, int[] idx) {
+    final int len = adapter.size(data);
+    MwpIndex[] I = (MwpIndex[]) Array.newInstance(MwpIndex.class, len);
+
+    int j = 0;
+    long correction = 0;
+    while(j < len) {
+      int k = j;
+      int t = 1;
+      double adjust = 0.0;
+
+      while((k < len - 1) && (adapter.getDouble(data, idx[k]) == adapter.getDouble(data, idx[k + 1]))) {
+        adjust += k;
+        k++;
+        t++;
+      }
+
+      if(k > j) {
+        double adjusted = (adjust + k) / t;
+        correction += (t * t * t) - t;
+
+        for(int m = j; m <= k; m++) {
+          I[m] = new MwpIndex(idx[m], adjusted, correction);
+        }
+      }
+      else
+        I[j] = new MwpIndex(idx[j], j, correction);
+      j += t;
+    }
+
+    return I;
+  }
+
+  /**
+   * Efficient implementation of MWP statistical test using appropriate index structure as described in Algorithm 3
+   * of reference paper. Tailored to MCDE Framework.
+   *
+   * @param start           Starting index value for statistical test
+   * @param end             End index value for statistical test
+   * @param slice           Return value of randomSlice() created with the index that is not for the reference dimension
+   * @param corrected_ranks Index of the reference dimension, return value of corrected_ranks() computed for reference dimension
+   * @return p-value from two sided Mann-Whitney-U test
+   */
+  public double statistical_test(int start, int end, boolean[] slice, MwpIndex[] corrected_ranks) {
+    double R = 0.0;
+    long n1 = 0;
+    for(int j = start; j < end; j++) {
+
+      if(slice[corrected_ranks[j].index]) {
+        R += corrected_ranks[j].adjusted;
+        n1++;
+      }
+    }
+
+    // This is to cancel the offset in case the marginal restriction does not start from 0
+    // see "acc - (cutStart * count)" is reference implementation of MWP
+    R -= start * n1;
+
+    final int cutLength = end - start;
+    if((n1 == 0) || (n1 == cutLength)) {
+      return 1;
+    }
+
+    final double U = R - ((double) (n1 * (n1 - 1))) / 2;
+    final long n2 = cutLength - n1;
+
+    if(n1 + n2 > two_times_sqrt_max_long) {
+      throw new AbortException("Long type overflowed. Too many objects: Please subsample and try again with smaller data set.");
+    }
+
+    final long b_end = corrected_ranks[(end - 1)].correction;
+    final long b_start = start == 0 ? 0 : corrected_ranks[(start - 1)].correction;
+    final double correction = (double) (b_end - b_start) / (cutLength * (cutLength - 1));
+    final double std = Math.sqrt((((double) (n1 * n2)) / 12) * (cutLength + 1 - correction));
+
+    if(std == 0) {
+      return 0;
+    }
+    else {
```

Comment:
The `else {` + `}` is not necessary, because of the `return` above for this special case.